File: network systems and services laboratory/Práctica 1.1/Python Version/UDP_servidor.py
import socket

def create_socket(ip_version):
    """Crea un socket compatible con IPv4 o IPv6 según la dirección IP."""
    try:
        if ':' in ip_version:
            return socket.socket(socket.AF_INET6, socket.SOCK_DGRAM)
        else:
            return socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    except socket.error as e:
        print(f"Error al crear el socket: {e}")
        return None

def udp_server(host, port):
    server_socket = create_socket(host)
    if not server_socket:
        return
    try:
        server_socket.bind((host, port))
        print(f"Servidor UDP escuchando en {host}:{port}")
        while True:
            try:
                data, addr = server_socket.recvfrom(1024)
                print(f"Mensaje recibido de {addr}: {data.decode()}")
                server_socket.sendto(data, addr)  # Eco de respuesta
            except socket.timeout:
                print("Tiempo de espera agotado esperando datos.")
            except Exception as e:
                print(f"Error al recibir datos: {e}")
    except Exception as e:
        print(f"Error en el servidor: {e}")
    finally:
        server_socket.close()

if __name__ == "__main__":
    import sys
    if len(sys.argv) < 3:
        print("Uso: python udp_server.py <IP> <Puerto>")
    else:
        ip, port = sys.argv[1], int(sys.argv[2])
        udp_server(ip, port)

File: network systems and services laboratory/Práctica 1.1/C Version/UDP_servidor.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <errno.h>

#define BUFFER_SIZE 1024

typedef struct sockaddr_in6 sockaddr_in6;
typedef struct sockaddr_in sockaddr_in;

typedef struct sockaddr sockaddr;

int create_socket(const char *ip) {
    int sock;
    if (strchr(ip, ':')) { // IPv6
        sock = socket(AF_INET6, SOCK_DGRAM, 0);
    } else { // IPv4
        sock = socket(AF_INET, SOCK_DGRAM, 0);
    }
    if (sock < 0) {
        perror("Error al crear el socket");
        return -1;
    }
    return sock;
}

void udp_server(const char *ip, int port) {
    int sock = create_socket(ip);
    if (sock < 0) return;
    
    sockaddr_in6 server_addr6;
    sockaddr_in server_addr;
    sockaddr *server_ptr;
    socklen_t addr_len;
    
    if (strchr(ip, ':')) { // IPv6
        memset(&server_addr6, 0, sizeof(server_addr6));
        server_addr6.sin6_family = AF_INET6;
        server_addr6.sin6_port = htons(port);
        server_addr6.sin6_addr = in6addr_any;
        server_ptr = (sockaddr *)&server_addr6;
        addr_len = sizeof(server_addr6);
    } else { // IPv4
        memset(&server_addr, 0, sizeof(server_addr));
        server_addr.sin_family = AF_INET;
        server_addr.sin_port = htons(port);
        server_addr.sin_addr.s_addr = INADDR_ANY;
        server_ptr = (sockaddr *)&server_addr;
        addr_len = sizeof(server_addr);
    }
    
    if (bind(sock, server_ptr, addr_len) < 0) {
        perror("Error en bind");
        close(sock);
        return;
    }
    
    printf("Servidor UDP escuchando en %s:%d\n", ip, port);
    char buffer[BUFFER_SIZE];
    sockaddr_in6 client_addr;
    socklen_t client_len = sizeof(client_addr);
    
    while (1) {
        ssize_t recv_len = recvfrom(sock, buffer, BUFFER_SIZE, 0, (sockaddr *)&client_addr, &client_len);
        if (recv_len < 0) {
            perror("Error al recibir datos");
            continue;
        }
        buffer[recv_len] = '\0';
        printf("Mensaje recibido: %s\n", buffer);
        sendto(sock, buffer, recv_len, 0, (sockaddr *)&client_addr, client_len);
    }
    close(sock);
}

int main(int argc, char *argv[]) {
    if (argc < 3) {
        printf("Uso: %s <IP> <Puerto>\n", argv[0]);
        return 1;
    }
    udp_server(argv[1], atoi(argv[2]));
    return 0;
}
